core/src/execution/datafusion/expressions/binary.rs

```diff
@@ -0,0 +1,202 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use std::any::Any;
+use std::fmt::{Display, Formatter};
+use std::hash::Hasher;
+use std::sync::Arc;
+
+use arrow_array::{BooleanArray, RecordBatch};
+use arrow_schema::{DataType, Schema};
+use datafusion_common::{DataFusionError, Result, ScalarValue};
+use datafusion_expr::interval_arithmetic::Interval;
+use datafusion_expr::sort_properties::ExprProperties;
+use datafusion_expr::{ColumnarValue, Operator};
+use datafusion_physical_expr::expressions::{BinaryExpr, Literal};
+use datafusion_physical_expr_common::physical_expr::{down_cast_any_ref, PhysicalExpr};
+
+use crate::execution::datafusion::expressions::EvalMode;
+
+use super::arithmetic_overflow_error;
+
+#[derive(Debug, Hash, Clone)]
+pub struct CometBinaryExpr {
+    left: Arc<dyn PhysicalExpr>,
+    op: Operator,
+    right: Arc<dyn PhysicalExpr>,
+    eval_mode: EvalMode,
+    inner: Arc<BinaryExpr>,
+}
+
+impl CometBinaryExpr {
+    pub fn new(
+        left: Arc<dyn PhysicalExpr>,
+        op: Operator,
+        right: Arc<dyn PhysicalExpr>,
+        eval_mode: EvalMode,
+    ) -> Self {
+        Self {
+            left: Arc::clone(&left),
+            op,
+            right: Arc::clone(&right),
+            eval_mode,
+            inner: Arc::new(BinaryExpr::new(left, op, right)),
+        }
+    }
+
+    fn fail_on_overflow(&self, batch: &RecordBatch, result: &ColumnarValue) -> Result<()> {
+        if self.eval_mode == EvalMode::Ansi && self.op == Operator::Plus {
+            match result.data_type() {
+                DataType::Int8 | DataType::Int16 | DataType::Int32 | DataType::Int64 => {
+                    self.check_int_overflow(batch, result)?
+                }
+                _ => {}
+            }
+        }
+        Ok(())
+    }
+
+    fn check_int_overflow(&self, batch: &RecordBatch, result: &ColumnarValue) -> Result<()> {
+        let check_overflow_expr = Arc::new(BinaryExpr::new(
```

Comment:
Since arrow provide overflow checked kernels (https://github.com/apache/arrow-rs/pull/2643)  does it makes sense to use those directly rather than re-implementing it?

Reply:
I'm going to review it. Thank you!

Reply:
Hi @eejbyfeldt , I've been looking at datafusion and I don't see any options to use those arrow operations from datafusion physical expressions. Do you know if this is implemented yet?

Reply:
@planga82  Can we make use of this arithmetic kernel https://docs.rs/arrow/latest/arrow/compute/kernels/numeric/fn.add.html to compute the addition and throw error based on the result.

Reply:
Can we do something like this ?
```rust
    fn evaluate(&self, batch: &RecordBatch) -> Result<ColumnarValue> {
        use arrow::compute::kernels::numeric::add;
        let lhs = self.left.evaluate(batch)?;
        let rhs = self.left.evaluate(batch)?;
        match (self.op, self.eval_mode) {
            (Operator::Minus, EvalMode::Ansi)  => {
              return apply(&lhs, &rhs, add)
            },
            _ => {
                self.inner.evaluate(batch)
            }
        }
    }
```

But the visibility of `apply` fn is more restrcited `pub (crate)`. We might need to raise a PR with datafusion to make it public

Reply:
My concern is that if we directly use the kernels to perform the operations instead of reusing the physical datafusion expression we may lose functionality or have to reimplement it here.

Reply:
Well put. I agree to what you are saying. I was thinking to override the implementation but your thoughts makes much more sense, safer & cleaner.

Reply:
I agree with @eejbyfeldt that we should just use the existing `add_checked` or `add_wrapped` kernels in arrow-rs that already provide the functionality that we need (unless we discover any compatibility issue compared to the JVM addExact logic). I will create an example to demonstrate how to use this and will post here later today